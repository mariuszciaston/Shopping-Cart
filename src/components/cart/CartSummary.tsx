import { Button } from "@/components/ui/Button";
import { ShoppingCart } from "@/types/types";

function CartSummary({ productsData, itemsInCart }: ShoppingCart) {
  return (
    <>
      <div className="flex justify-between">
        <div className="text-lg font-bold">Total &nbsp;</div>
        <div className="text-lg font-bold" data-testid="total-amount">
          $
          {itemsInCart
            .reduce(
              (acc, item) =>
                acc +
                item.quantity *
                  (productsData?.find((product) => product.id === item.id)
                    ?.price ?? 0),
              0,
            )
            .toFixed(2)}
        </div>
      </div>

      <Button
        className="mx-auto bg-blue-500 px-8 py-6 text-lg font-bold hover:bg-blue-500"
        variant="default"
        onClick={() => {
          if (itemsInCart.length > 0) {
            alert(`Congratulations! You successfully bought the products.`);
          } else {
            alert("Cart is empty. Please add items before checking out.");
          }
        }}
      >
        Checkout
      </Button>
    </>
  );
}

export default CartSummary;
