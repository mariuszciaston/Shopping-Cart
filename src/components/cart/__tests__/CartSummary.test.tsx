import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import CartSummary from "../CartSummary";
import { describe, expect, it, vi } from "vitest";

describe("CartSummary component", () => {
  const productsData = [
    {
      id: 1,
      price: 10.0,
      image: "image1.jpg",
      title: "title1",
      description: "description1",
    },
    {
      id: 2,
      price: 20.0,
      image: "image2.jpg",
      title: "title2",
      description: "description2",
    },
    {
      id: 3,
      price: 0.44,
      image: "image3.jpg",
      title: "title3",
      description: "description3",
    },
  ];

  const itemsInCart = [
    { id: 1, quantity: 2 },
    { id: 2, quantity: 1 },
    { id: 3, quantity: 1 },
  ];

  it("renders CartSummary ", () => {
    const { container } = render(
      <CartSummary productsData={productsData} itemsInCart={itemsInCart} />,
    );

    expect(container).toBeInTheDocument();
  });

  it("calculates total amount", () => {
    render(
      <CartSummary productsData={productsData} itemsInCart={itemsInCart} />,
    );

    const totalAmount = screen.getByTestId("total-amount");
    expect(totalAmount).toHaveTextContent("$40.44");
  });

  it("calculates total amount without items in cart", () => {
    const productsData: [] = [];
    const itemsInCart: [] = [];

    render(
      <CartSummary productsData={productsData} itemsInCart={itemsInCart} />,
    );

    const totalAmount = screen.getByTestId("total-amount");
    expect(totalAmount).toHaveTextContent("$0");
  });

  it("shows unsuccessful alert when checkout button is clicked with an empty cart", () => {
    const productsData: [] = [];
    const itemsInCart: [] = [];

    render(
      <CartSummary productsData={productsData} itemsInCart={itemsInCart} />,
    );

    window.alert = vi.fn();

    const checkoutButton = screen.getByRole("button", { name: "Checkout" });
    fireEvent.click(checkoutButton);

    expect(window.alert).toHaveBeenCalledWith(
      "Cart is empty. Please add items before checking out.",
    );
  });

  it("shows successful alert when checkout button is clicked with items in cart", () => {
    render(
      <CartSummary productsData={productsData} itemsInCart={itemsInCart} />,
    );

    window.alert = vi.fn();

    const checkoutButton = screen.getByRole("button", { name: "Checkout" });
    fireEvent.click(checkoutButton);

    expect(window.alert).toHaveBeenCalledWith(
      "Congratulations! You successfully bought the products.",
    );
  });
});
