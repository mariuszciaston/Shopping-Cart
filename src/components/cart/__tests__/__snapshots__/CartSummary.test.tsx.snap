// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartSummary component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex justify-between"
  >
    <div
      class="text-lg font-bold"
    >
      Total Â 
    </div>
    <div
      class="text-lg font-bold"
      data-testid="total-amount"
    >
      $40.44
    </div>
  </div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 text-primary-foreground shadow h-9 mx-auto bg-blue-500 px-8 py-6 text-lg font-bold hover:bg-blue-500"
  >
    Checkout
  </button>
</DocumentFragment>
`;
